src/hooks/useShapefileUpload.ts

```diff
@@ -0,0 +1,193 @@
+
+const useShapefileUpload = (options: UseShapefileUploadOptions): UseShapefileUploadResult => {
+  const { allowedGeometryTypes = DEFAULT_ALLOWED_TYPES, onSuccess } = options;
+
+  const [error, setError] = useState<string | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+
+  const handleFileChange = useCallback(
+    async (event: React.ChangeEvent<HTMLInputElement>) => {
+      const file = event.target.files?.[0];
+
+      if (!file) {
+        return;
+      }
+
+      if (!file.name.toLowerCase().endsWith('.zip')) {
+        setError(ZIP_FILE_ERROR);
+        event.target.value = '';
+        return;
+      }
+
+      setIsLoading(true);
+      setError(null);
+
+      try {
+        const arrayBuffer = await file.arrayBuffer();
+        const geojson = (await shp(arrayBuffer)) as FeatureCollectionWithCrs;
+
+        if (!geojson?.features?.length) {
+          throw new Error(EMPTY_FILE_ERROR);
+        }
+
+        const sourceSpatialReference = getSpatialReferenceFromGeoJSON(geojson);
+        const arcgisFeatures = geojsonToArcGIS(geojson) as Array<{ geometry: ReturnType<__esri.Geometry['toJSON']> }>;
+
+        if (!projectOperator.isLoaded()) {
+          await projectOperator.load();
+        }
+        const esriGeometries = arcgisFeatures.reduce<__esri.Geometry[]>((acc, feature) => {
+          const geometry = geometryFromJSON(feature.geometry);
+          if (geometry) {
+            ensureSpatialReference(geometry, sourceSpatialReference);
+            if (geometry.spatialReference?.wkid === WEB_MERCATOR.wkid) {
+              acc.push(geometry);
+            }
+
+            const projected = projectOperator.execute(geometry, WEB_MERCATOR);
+
+            if (!projected) {
+              throw new Error('Unable to project shapefile geometry to Web Mercator (EPSG:3857).');
+            }
+
+            acc.push(projected);
+          }
+
+          return acc;
+        }, []);
+
+        if (!esriGeometries.length) {
+          throw new Error(EMPTY_FILE_ERROR);
+        }
+
+        const geometryType = (esriGeometries[0]?.type ?? 'polygon') as AllowedGeometryType;
+
+        if (!allowedGeometryTypes.includes(geometryType)) {
+          throw new Error(UNSUPPORTED_GEOMETRY_ERROR);
+        }
+
+        let unionedGeometry: __esri.Geometry | nullish = null;
+        const firstProjectedGeometry = esriGeometries[0];
+
+        if (!firstProjectedGeometry) {
+          throw new Error(EMPTY_FILE_ERROR);
+        }
+
+        if (geometryType === 'point' || geometryType === 'multipoint') {
+          unionedGeometry = combinePoints(esriGeometries);
+        } else if (geometryType === 'polygon') {
+          unionedGeometry =
+            esriGeometries.length === 1
+              ? firstProjectedGeometry
+              : (unionOperator.executeMany(esriGeometries as __esri.GeometryUnion[]) as __esri.Geometry);
+        } else {
+          // For polylines, unionOperator is not applicable; use first geometry
```

Comment:
is there an issue with having many paths? I would assume they can be union'd like the other geometry types?